### ISSUE TITLE
Custom option type factory is bound to a folder

### ISSUE THREAD

**Issue by tzyganu** [magento/magento2#24726]
### Summary (*)
Method `Magento\Catalog\Model\Product\Option::groupFactory` binds all accessory types with to the `Magento\Catalog\Model\Product\Option\Type` namespace.

### Examples (*)
Let's say I want to create a new type of custom option (let's call it `Accessory`).  
This will be rendered different and will behave differently from the other option types (select, text, ...)
So I need another group type and another class to handle my code.  
The method `Magento\Catalog\Model\Product\Option::groupFactory` looks like this.

```
    public function groupFactory($type)
    {
        $group = $this->getGroupByType($type);
        if (!empty($group)) {
            return $this->optionTypeFactory->create(
                'Magento\Catalog\Model\Product\Option\Type\\' . $this->string->upperCaseWords($group)
            );
        }
        throw new LocalizedException(__('The option type to get group instance is incorrect.'));
    }
````
This means that the class that needs to be instantiated must be located in the namespace `Magento\Catalog\Model\Product\Option\Type\`. 
Not cool.  
I know that this can be overcome through a n around plugin or a preference for `Magento\Catalog\Model\Product\Option\Type\Accessory` in my case, but this is not optimal.  
Ideally it should work out of the box.  

### Proposed solution
Have an array for mapping type to class name.  
This array can be an optional constructor parameter so custom modules can add elements to it.  
and the `groupFactory` method should look in this array first, if nothing is there, fallback to current way of instantiating classes.

Bonus:
Have a configurable array for group to type that is now hardcoded in `getGroupByType` in the same class.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-20529` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by zhartaunik** [comment]
The issue is still open after the PR merge. Does it require sth from my side?

**Comment by magento-engcom-team** [comment]
Hi @tzyganu. Thank you for your report. 
The issue has been fixed in magento/magento2#25055 by @zhartaunik in 2.3-develop branch
Related commit(s):
 - [de97b0ae727ed009ac2707e4be979366c565a333](https://github.com/magento/magento2/commit/de97b0ae727ed009ac2707e4be979366c565a333)
 - [4a1f3ab77f8d0dbbb7fc4d8c6dbb6ec70c400e74](https://github.com/magento/magento2/commit/4a1f3ab77f8d0dbbb7fc4d8c6dbb6ec70c400e74)
 - [b5692d93b4fdae479a33ec698d92ff48359f81a9](https://github.com/magento/magento2/commit/b5692d93b4fdae479a33ec698d92ff48359f81a9)
 - [1109bbd6a6fa2e0e2ef27997dbca83f0e0178f98](https://github.com/magento/magento2/commit/1109bbd6a6fa2e0e2ef27997dbca83f0e0178f98)
 - [600c9a8d602ff00b3a883f04d109629fd7d51bf2](https://github.com/magento/magento2/commit/600c9a8d602ff00b3a883f04d109629fd7d51bf2)
 - [de29ee24df829e8dc6e3bc431bd753e693517d04](https://github.com/magento/magento2/commit/de29ee24df829e8dc6e3bc431bd753e693517d04)
 - [20b0f26d66a7dad6cd87f64c88f033fce2254347](https://github.com/magento/magento2/commit/20b0f26d66a7dad6cd87f64c88f033fce2254347)
 - [90ca5f0803c827d7484d9c239f10ff3de1b09030](https://github.com/magento/magento2/commit/90ca5f0803c827d7484d9c239f10ff3de1b09030)
 - [de9ca56b01c7c8cf4805d189fc6c6052a8491a38](https://github.com/magento/magento2/commit/de9ca56b01c7c8cf4805d189fc6c6052a8491a38)

The fix will be available with the upcoming 2.3.4 release.
